controllers/openstackmachine_controller.go

```diff
@@ -583,7 +583,28 @@ func (r *OpenStackMachineReconciler) getOrCreateMachineServer(ctx context.Contex
 			}
 			return openStackCluster.Spec.IdentityRef
 		}()
-		machineServerSpec := openStackMachineSpecToOpenStackServerSpec(&openStackMachine.Spec, identityRef, compute.InstanceTags(&openStackMachine.Spec, openStackCluster), failureDomain, userDataRef, getManagedSecurityGroup(openStackCluster, machine), openStackCluster.Status.Network.ID)
+		// If user has provided .spec.ports (non-empty), skip .Status.Network entirely
+		// Otherwise, fall back to openStackCluster.Status.Network.ID
+		// This supports HCP - Hosted Control Plane usage
+		defaultNetworkID := ""
+		if len(openStackMachine.Spec.Ports) == 0 {
+			if openStackCluster.Status.Network == nil || openStackCluster.Status.Network.ID == "" {
+				return nil, fmt.Errorf(
```

Comment:
Please can you wrap this in a [TerminalError](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/c14cb8c029f55a1b620942b27b82b4c38c59f4dc/pkg/utils/errors/terminal.go#L34) so we'll stop trying to reconcile it?

Reply:
I believe the OpenStackServer controller has a great example.